### PR TITLE
make json serialization overridable for events

### PR DESCRIPTION
**Pull request by nathaniel-may**
### Description

This PR does three things:

1. Instead of inspecting the Python dictionary values for scrubability before converting to a string, convert the whole log line to a json string and scrub the string.

which requires

2. More graceful handling of events with non json serializable field types, by making the serialization function overridable.

which implies the need for

3. A test that attempts to serialize one instance of each concrete event type into JSON. This is not a perfect test becuase some of these events are annotated with the `Any` type so mypy nor this test will catch if some instances of that class that we make somewhere in the code base are not serializable. However, in that case instead of exploding or hanging, users will see something like `{ ... "some_key": "JSON_SERIALIZE_FAILED: MyClass", ... }`.

### Checklist

- [ ] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [ ] I have run this code in development and it appears to resolve the stated issue
- [ ] This PR includes tests, or tests are not required/relevant for this PR
- [ ] I have updated the `CHANGELOG.md` and added information about my change

### CODE REVIEW

**Comment by emmyoop** [comment]
Can you add the `test all` label?

**Comment by nathaniel-may** [comment]
> `TypeError: Object of type VersionSpecifier is not JSON serializable`
ahh, I see why you caught `TypeError`. I'm going to try and fix this with a different approach and see what you think.

File: core/dbt/events/functions.py [hunk 1]
```diff
@@ -181,17 +153,8 @@ def create_json_log_line(e: T_Event, msg_fn: Callable[[T_Event], str]) -> str:
...
-        log_line = json.dumps(scrub_dict_secrets(values), sort_keys=True)
-    except TypeError:
-        # the only key currently throwing errors is 'data'.  Expand this list
-        # as needed if new issues pop up
-        values['data'] = None
-        values['event_data_serialized'] = False
-        log_line = json.dumps(scrub_dict_secrets(values), sort_keys=True)
-    return log_line
+    raw_log_line = json.dumps(values, sort_keys=True)
+    return scrub_secrets(raw_log_line, env_secrets())
```
Comment (emmyoop) [inline core/dbt/events/functions.py:157]: I think you may still need the try statement to build the raw log line.  Then scrub on the return.  The `data` field has unpredictable contents.

Comment (nathaniel-may) [reply core/dbt/events/functions.py:157]: The data field has very predictable contents! We wrote them all down in `types.py`. I think we just have to handle them better. I'll propose something here and see what you think.

**Comment by nathaniel-may** [comment]
having trouble debugging the test failures here. I'll try again in the morning.

**Comment by nathaniel-may** [comment]
@emmyoop helped me figure out that the hanging tests actually error out on a local `dbt run` which was helpful. Now that I've fixed the root causes of those exceptions, both the tests and `dbt run` hang. I'll have to come back to this later.

**Comment by nathaniel-may** [comment]
hooray! Just the windows tests fail. Adding a test for serialization now.

**Comment by leahwicz** [comment]
The changes LGTM but the Adapter tests seem to be failing so I'm concerned about that

**Comment by gshank** [comment]
In the rest of our code we use Mashumaro for serialization ('to_dict' is usually mashumaro). It's too late to use it for this now, but we might want to look at the advantages/disadvantages for that later.

**Comment by nathaniel-may** [comment]
@leahwicz the adapter tests are only the windows ones which is a known issue.

@gshank good call we should definitely do that.

**Review by emmyoop — APPROVED** [review]

**Review by leahwicz — APPROVED** [review]

**Pull request merged**